sentry-types/src/protocol/envelope.rs

```diff
@@ -236,6 +302,125 @@ impl Envelope {
 
         Ok(())
     }
+
+    /// Creates a new Envelope from slice.
+    pub fn from_slice(slice: &[u8]) -> Result<Envelope, EnvelopeError> {
+        let (header, offset) = Self::parse_header(slice)?;
+        let items = Self::parse_items(slice, offset)?;
+
+        let mut envelope = Envelope {
+            event_id: header.event_id,
+            ..Default::default()
+        };
+
+        for item in items {
+            envelope.add_item(item);
+        }
+
+        Ok(envelope)
+    }
+
+    /// Creates a new Envelope from path.
+    pub fn from_path<P: AsRef<Path>>(path: P) -> Result<Envelope, EnvelopeError> {
+        let bytes = std::fs::read(path).map_err(|_| EnvelopeError::UnexpectedEof)?;
+        Envelope::from_slice(&bytes)
+    }
+
+    fn parse_header(slice: &[u8]) -> Result<(EnvelopeHeader, usize), EnvelopeError> {
+        let mut stream = serde_json::Deserializer::from_slice(slice).into_iter();
+
+        let header: EnvelopeHeader = match stream.next() {
+            None => return Err(EnvelopeError::MissingHeader),
+            Some(Err(error)) => return Err(EnvelopeError::InvalidHeader(error)),
+            Some(Ok(header)) => header,
+        };
+
+        // Each header is terminated by a UNIX newline.
+        Self::require_termination(slice, stream.byte_offset())?;
+
+        Ok((header, stream.byte_offset() + 1))
+    }
+
+    fn parse_items(slice: &[u8], mut offset: usize) -> Result<Vec<EnvelopeItem>, EnvelopeError> {
+        let mut items = Vec::new();
+
+        while offset < slice.len() {
```

Comment:
Instead of having a `mut offset` and sub-slicing from the original `slice` all the time, we can just have `mut slice` and do a `slice = slice.get(offset..)?` after we read each item, and above in `from_slice`.
But it doesn’t really matter that much anyway.